File: lunarrx-core/src/main/java/com/cisco/vss/lunar/rx/plugin/core/LunarTrackItemStreamTransformer.java
package com.cisco.vss.lunar.rx.plugin.core;

import rx.Observable;
import static com.cisco.vss.rx.java.Conversions.*;
import rx.functions.Func1;

public class LunarTrackItemStreamTransformer<T extends LunarTrackItem, R extends LunarTrackItem> implements Func1<Observable<byte[]>, Observable<? extends R>> {
	private final Class<T>                                      sourceType;
	private final Func1<Observable<T>, Observable<? extends R>> transform;
	
	protected LunarTrackItemStreamTransformer(final Class<T> sourceType, Func1<Observable<T>, Observable<? extends R>> transform) {
		this.sourceType = sourceType;
		this.transform  = transform;
	}

	@Override
	public Observable<? extends R> call(final Observable<byte[]> input) {
		return this.transform.call(
			input.map(byte2String).flatMap(jsonString2Object(sourceType))
		);
	}
}

File: lunarrx-scala/src/main/scala/com/cisco/vss/lunar/rx/plugin/scala/LunarTrackItemStreamTransformer.scala
package com.cisco.vss.lunar.rx.plugin.scala

import com.cisco.vss.lunar.rx.plugin._
import rx.lang.scala.JavaConversions._
import rx.lang.scala._
import com.cisco.vss.lunar.rx.plugin.core.TrackItem

abstract class LunarTrackItemStreamTransformer[T <: TrackItem, R <: TrackItem] (lunar: Lunar)(implicit sourceType: Class[T], resultType: Class[R]){
  private [scala] val asJava: core.LunarTrackItemStreamTransformer[T, R] = new core.LunarTrackItemStreamTransformer[T, R](lunar.asJavaLunar, sourceType, resultType) {
    @Override
    def transformT(input: rx.Observable[T]) : rx.Observable[_ <: R] = {
        val result     = transformA(toScalaObservable(input))
        val javaResult = toJavaObservable(result) 
        return javaResult
    }
  } 

  //Name?!!!!
  def transformA(input: Observable[T]): Observable[R]
}